src/Utilities/Tuple.hpp

```diff
@@ -174,3 +174,37 @@ constexpr inline void tuple_transform(
       tuple, std::forward<N_aryOp>(op),
       std::make_index_sequence<sizeof...(Elements)>{}, args...);
 }
+
+namespace tuple_impl_detail {
+
+template <typename Invokable, typename... Args, size_t... ArgsIndices>
+auto apply_impl(const std::tuple<Args...>& args, Invokable&& invokable,
+                std::index_sequence<ArgsIndices...> /* meta */) {
+  return invokable(std::get<ArgsIndices>(args)...);
+}
+
+}  // namespace tuple_impl_detail
+
+namespace cpp17 {
+
+/*!
+ * \ingroup UtilitiesGroup
+ * \brief Invoke the `invokable` with the `args` expanded in a parameter pack
+ *
+ * Note that this function is not quite as general as C++17's `std::apply`.
+ *
+ * Here is an example how to use the function:
+ *
+ * \snippet Test_Tuple.cpp expand_tuple_example
+ *
+ * This is the function being called in the above example:
+ *
+ * \snippet Test_Tuple.cpp expand_tuple_example_function
+ */
+template <typename Invokable, typename... Args>
+auto apply(Invokable&& invokable, const std::tuple<Args...>& args) {
```

Comment:
I think it would be better to take the tuple by rvalue reference, and to `std::forward` when passing to the impl, and do the same rvalue reference and forwarding in the implementation.
I see the "possible implementation" on cppreference uses a cpp20 feature `invoke`... however, the main complexities of `invoke` involve handling member functions and member objects, so I think the best way to go about this is to just assume for the moment that we're not applying member function pointers and use the simple version in the `apply_impl`:
```
return std::forward<Invokable>(invokable)(std::forward<Args>(args)...)
```
and make the return type of the functions `decltype(auto)` and the functions `constexpr` in case the invokable is sufficiently simple to be brought in at compile-time.

Reply:
agreed, I pushed a fixup a few minutes ago that does pretty much that :D

Reply:
do I need the `std::forward<Invokable>(invokable)` at the beginning? I'm not quite sure what that's for

Reply:
I think it is best to do that for funny corner cases. My understanding is that it ensures appropriate forwarding of the invokable as reference. So for instance, if passing a non-const lvalue reference to a class with a call operator that alters the class member variables, I think it would appropriately ensure that the class members get altered; and call the right version of overloaded call operators when deciding between const and non-const versions according to whether the invokable was passed as const or non-const reference.